docs/specs/lcb-signals-and-events.md

```diff
@@ -0,0 +1,222 @@
+
+2. **Dispatch**: do work in response to the event
+
+3. **Leave**: perform any tasks that should be carried out after event handling (for example, tearing down any state set up during event enter)
+
+Before beginning to process an event, the event manager first determines which control should be the **target** of the event (taking into account visibilities, and bounding rectangles) [2].
+
+A list of "interested controls" is built.  This recursively contains every parent of the target.  This is used to determine which objects should have their event handlers run.
+
+**Notes:**
+
+1. The enter phase and leave phase facilitate the implementation of behaviour that wraps normal event dispatch.  For example, consider a composed widget with a border that highlights when the mouse is inside it.  To ensure that highlight changes sequence correctly with with its child widgets, the highlight must be enabled during the `MouseEnter` event's enter phase, and hidden during the `MouseLeave` event's leave phase.
+
+2. How the target of an event is determined may vary a lot on a per-event-type basis (for example, the target for a mouse click will depend on the mouse position, but the target for a keyboard event may depend on the keyboard focus).  For the purposes of this document, assume that it is an arbitrary process.
+
+### Event data
+
+When an event is submitted to the event manager, it has a **event name** and a set of **event data**.  The event data is an array containing all of the captured data relating to the event itself [1].
+
+In an event handler, `the event` evaluates to the event data array [2].
+
+Evaluating `the event` outwith an event handler is an error.
+
+`the event` is immutable.
+
+The `the event target` expression evaluates to the event's target control [3].
+
+**Notes:**
+
+1. For example, when creating a mouse event, it is important to capture the position of the mouse at the time the event began to be dispatched into the event data.  If computationally-intensive event handlers are run, the user could move the mouse between event creation and event handling.
+
+2. This permits simple implementation of syntactic sugar for frequently-needed event data.  For example:
+
+       the mouse position of the event
+
+   can easily be implemented as:
+
+       the event["mouse position"]
+
+
+3. It might be useful to provide `the event target is me` (or equivalent syntax) to test whether an event is targeted at the handling control or one of its children.
+
+
+### Event handlers
+
+All event handlers (for enter, dispatch and leave phases) must conform to:
+
+    public type EventHandler() returns any
+
+The return value of all event handlers is discarded.
+
+### Enter phase
+
+During the enter phase, **pre-event** handlers are run.  They are run from "outside-in"; a parent widget's pre-event handler runs before the pre-event handler of its child widgets.
```

Comment:
I did a quick test on a behavior hierarchy and found before handlers are evaluated in child -> parent order. In order for my suggestion to be a good one I think we would need to be prepared to change that to child <- parent order and execute all before handlers in the hierarchy even if the before handler is unpassed. I actually think this would be more useful for behavior scripts if it were feasible to make the change. I suspect it is unlikely that huge numbers of people would be impacted by the change as it would only impact those that have overridden a parent behavior before handler with a child one.